File: ocaml/tennis3.ml

let show_score p1 p2 =
  let p1N = "player1" in
  let p2N = "player2" in
  if (p1 < 4 && p2 < 4) && (p1 + p2 < 6) then
  begin
    let p = [| "Love"; "Fifteen"; "Thirty"; "Forty" |] in
    let s = p.(p1) in
    if (p1 = p2) then s ^ "-All" else s ^ "-" ^ p.(p2)
  end else begin
    if p1 = p2 then
      "Deuce"
    else
    begin
      let s = if p1 > p2 then p1N else p2N in
      if (p1-p2) * (p1-p2) = 1 then "Advantage " ^ s else "Win for " ^ s
    end
  end

File: erlang/src/tennis3.erl
-module(tennis3).

-include("tennis3.hrl").
-export([score/1]).

-spec score(#game{}) -> string().
score(#game{player1 = P1, player2 = P2, player1_name = P1N, player2_name = P2N}) ->
  case P1 < 4 andalso P2 < 4 andalso not(P1 + P2 == 6) of
    true ->
      P = ["Love", "Fifteen", "Thirty", "Forty"],
      S = lists:nth(P1 + 1, P),
      case P1 == P2 of
        true ->
          S ++ "-All";
        false ->
          S ++ "-" ++ lists:nth(P2 + 1, P)
      end;
    false ->
      case P1 == P2 of
        true -> "Deuce";
        false ->
          S = case P1 > P2 of
                true -> P1N;
                false -> P2N
              end,
          case (P1 - P2) * (P1 - P2) == 1 of
            true -> "Advantage " ++ S;
            false -> "Win for " ++ S
          end
      end
  end.
